File: go/tencent/backtrack/subsets.go
package main

func subsets(nums []int) [][]int {
	res := make([][]int, 0)
	recursionSub(&res, []int{}, nums, 0)
	return res
}

func recursionSub(res *[][]int, temp []int, nums []int, start int) {
	cp := make([]int, len(temp))
	copy(cp, temp)
	*res = append(*res, cp)
	for i := start; i < len(nums); i++ {
		temp = append(temp, nums[i])
		recursionSub(res, temp, nums, i+1)
		temp = temp[:len(temp)-1]
	}
}

File: java/leetcode-explore/tencent/dp/Subsets.java
package tencent.dp;

import java.util.ArrayList;
import java.util.List;

public class Subsets {

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        recursion(res, new ArrayList<>(), 0, nums);
        return res;
    }

    public void recursion(List<List<Integer>> res, List<Integer> tmp, int start, int[] nums) {
        res.add(new ArrayList<>(tmp));
        for (int i = start; i < nums.length; i++) {
            tmp.add(nums[i]);
            recursion(res, tmp, i + 1, nums);
            tmp.remove(tmp.size() - 1);
        }
    }

    public static void main(String[] args) {
        System.out.println(new Subsets().subsets(new int[]{1,2,3}));
    }
}
